### PR TITLE
qemu/tests/cfg: Add new scenario for multi_host

### PR DESCRIPTION
**Pull request by spcui**
Do nic driver unload/load for migration.

Signed-off-by: Shuping Cui <scui@redhat.com>

ID: 1266315 1266316

### CODE REVIEW

File: qemu/tests/cfg/multi_host.cfg [hunk 1]
```diff
@@ -315,6 +315,41 @@
...
+                    iothreads = iothread0
+                    blk_extra_params_image1 = "iothread=${iothreads}"
+                    drive_werror_image1 = stop
+                    drive_rerror_image1 = stop
+                    migration_bg_command = "fio -filename /dev/vda --runtime 600 --rw=randread -name=randread0 > /dev/null &"
+                    migration_bg_check_command = "pgrep fio"
+                    migration_bg_kill_command = "pkill -9 fio"
+
+                - with_module_reload:
+                    only Linux
```
Comment (suqinhuang) [inline qemu/tests/cfg/multi_host.cfg]: how about Windows ?

File: qemu/tests/cfg/multi_host.cfg [hunk 2]
```diff
@@ -315,6 +315,41 @@
...
+                    only Host_RHEL.m7
+                    only virtio_blk
+                    type = migration_multi_host
+                    preprocess_env = yes
+                    iothreads = iothread0
+                    blk_extra_params_image1 = "iothread=${iothreads}"
+                    drive_werror_image1 = stop
+                    drive_rerror_image1 = stop
+                    migration_bg_command = "fio -filename /dev/vda --runtime 600 --rw=randread -name=randread0 > /dev/null &"
```
Comment (suqinhuang) [inline qemu/tests/cfg/multi_host.cfg]: is fio installed in guest?

File: qemu/tests/cfg/multi_host.cfg [hunk 3]
```diff
@@ -315,6 +315,41 @@
...
+                    filesize = 100
+                    transfer_timeout = 600
+                    virtio_net:
+                        module_name = virtio_net
+                    e1000:
+                        module_name = e1000
+                    rtl8139:
+                        module_name = 8139cp
+                    migration_bg_command = "cd /tmp;"
+                    migration_bg_command += " echo "while true; do rmmod ${module_name}; modprobe ${module_name}; sleep 3; done" > test.sh;"
```
Comment (suqinhuang) [inline qemu/tests/cfg/multi_host.cfg]: @xutian could you review this patch,
The code will be weak, if just send the bg cmd without any handling, we don't if the script is running, what happen in guest while test run in guest.

Thanks

File: qemu/tests/cfg/multi_host.cfg (hunk 2 above)
Comment (spcui) [inline qemu/tests/cfg/multi_host.cfg]: yes, will add install fio in .ks file

File: qemu/tests/cfg/multi_host.cfg [hunk 4]
```diff
@@ -332,6 +332,41 @@
...
+                    serial_login = yes
+                    not_login_after_mig = yes
+                    post_sub_test = "file_transfer"
+                    filesize = 100
+                    transfer_timeout = 600
+                    virtio_net:
+                        module_name = virtio_net
+                    e1000:
+                        module_name = e1000
+                    rtl8139:
```
Comment (suqinhuang) [inline qemu/tests/cfg/multi_host.cfg]: pls move this part to guest-os

File: qemu/tests/cfg/multi_host.cfg [hunk 5]
```diff
@@ -332,6 +332,41 @@
...
+                        module_name = 8139cp
+                    migration_bg_command = "cd /tmp &&"
+                    migration_bg_command += " nohup python module_test.py ${module_name} > module_reload.out 2>&1 &"
+                    migration_bg_check_command = "pgrep python"
+                    migration_bg_kill_command = "pkill -9 python && modprobe ${module_name}"
```
Comment (suqinhuang) [inline qemu/tests/cfg/multi_host.cfg]: pls kill the pid

File: qemu/tests/cfg/multi_host.cfg (hunk 2 above)
Comment (suqinhuang) [inline qemu/tests/cfg/multi_host.cfg]: pls use autotest.fio

File: qemu/deps/module_reload/module_test.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import sys
+import commands
+import time
+
+module_name = sys.argv[1]
+cmd = "rmmod -f %s && modprobe %s" % (module_name, module_name)
+
```
Comment (suqinhuang) [inline qemu/deps/module_reload/module_test.py]: need to check model after rmmod/modprobe

File: qemu/tests/cfg/multi_host.cfg (hunk 1 above)
Comment (xutian) [inline qemu/tests/cfg/multi_host.cfg]: http://www.bluestop.org/fio/

File: qemu/deps/module_reload/module_test.py [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+while True:
+    s, o = commands.getstatusoutput(cmd)
+    if s != 0:
+        print o
+        break
```
Comment (suqinhuang) [inline qemu/deps/module_reload/module_test.py]: use raise instead of break

File: qemu/tests/cfg/multi_host.cfg [hunk 8]
```diff
@@ -332,6 +332,41 @@
...
+                    transfer_timeout = 600
+                    virtio_net:
+                        module_name = virtio_net
+                    e1000:
+                        module_name = e1000
+                    rtl8139:
+                        module_name = 8139cp
+                    migration_bg_command = "cd /tmp &&"
+                    migration_bg_command += " nohup python module_test.py ${module_name} > module_reload.out 2>&1 &"
+                    migration_bg_check_command = "pgrep python"
```
Comment (xutian) [inline qemu/tests/cfg/multi_host.cfg]: make log more clearly

File: qemu/tests/migration_multi_host.py [hunk 9]
```diff
@@ -1,8 +1,11 @@
 import logging
 import time
 import random
+import os
```
Comment (xutian) [inline qemu/tests/migration_multi_host.py]: Please update your modules import order.

**Comment by xutian** [comment]
Please enhance driver_load/unload test, then call that test directly

File: qemu/tests/cfg/multi_host.cfg [hunk 10]
```diff
@@ -332,6 +332,31 @@
...
 
+                - with_module_reload:
+                    type = migration_multi_host_with_stress
+                    preprocess_env = yes
+                    #serial_login = yes
+                    #not_login_after_mig = yes
+                    bg_stress_test = driver_load
+                    repeats = 20
+                    nics += " nic2"
+                    virtio_net:
```
Comment (xutian) [inline qemu/tests/cfg/multi_host.cfg]: move guest-hw related cfg to guest-hw.cfg.

thanks,
Xu

File: qemu/tests/cfg/multi_host.cfg [hunk 11]
```diff
@@ -332,6 +332,31 @@
...
+                        module_name = virtio_net
+                        nic_model_nic2 = e1000
+                    e1000:
+                        module_name = e1000
+                        nic_model_nic2 = virtio_net
+                    rtl8139:
+                        module_name = 8139cp
+                        nic_model_nic2 = virtio_net
+                    driver_id_pattern = ${module_name}
+                    driver_id_cmd = lsmod | grep ${module_name}
```
Comment (xutian) [inline qemu/tests/cfg/multi_host.cfg]: How about windows guest?

Comment (spcui) [inline qemu/tests/cfg/multi_host.cfg]: deps on: https://github.com/avocado-framework/avocado-vt/pull/295/files

**Comment by xutian** [comment]
looks good to me, merged

**Pull request merged**